Coverity fixes

pks-t opened the pull request:
Another round of Coverity fixes. Mostly fixes for memory leaks on error paths and some smallish error value fixes. Otherwise some changes such that Coverity won't complain anymore on wrong positives, especially for `git_vector_foreach`.

ethomson commented on src/path.c:
```diff
@@ -782,7 +782,8 @@ int git_path_resolve_relative(git_buf *path, size_t ceiling)
 int git_path_apply_relative(git_buf *target, const char *relpath)
 {
-	git_buf_joinpath(target, git_buf_cstr(target), relpath);
+	if (git_buf_joinpath(target, git_buf_cstr(target), relpath) < 0)
```
This should not be necessary.  `git_path_resolve_relative` [checks for an OOM on `target`](https://github.com/libgit2/libgit2/blob/594a5d12d4b9ba3ea542722805ada1028bd1b646/src/path.c#L708) as its first check...

pks-t replied:
Ah, okay, didn't check the other function indeed.

ethomson commented on src/xdiff/xmerge.c:
```diff
@@ -646,6 +646,8 @@ int xdl_merge(mmfile_t *orig, mmfile_t *mf1, mmfile_t *mf2,
 	    xdl_change_compact(&xe2.xdf2, &xe2.xdf1, xpp->flags) < 0 ||
 	    xdl_build_script(&xe2, &xscr2) < 0) {
+		xdl_free_script(xscr1);
+		xdl_free_env(&xe1);
```
We should upstream these changes to git itself.

pks-t replied:
Didn't realise this was imported from git. Will post to the ML tomorrow.

pks-t replied:
See [gmane](http://article.gmane.org/gmane.comp.version-control.git/287034) for the proposed patch.

ethomson commented on src/rebase.c:
```diff
@@ -270,8 +269,11 @@ static git_rebase *rebase_alloc(const git_rebase_options *rebase_opts)
-		if ((rebase->options.rewrite_notes_ref = git__strdup(rebase_opts->rewrite_notes_ref)) == NULL)
+		if ((rebase->options.rewrite_notes_ref =
+			git__strdup(rebase_opts->rewrite_notes_ref)) == NULL) {
+			git__free(rebase);
```
We're not in the habit of `free`ing things on allocation failure.  I don't like to have a different pattern here.  If the goal is to silence static code analysis warnings, then we could either:

1. Change this to return an `int` error code so that `GITERR_CHECK_ALLOC` works here, or
2. Silence this another way, perhaps with a macro like `GITERR_CHECK_ALLOC` that returns `NULL` instead of `-1`.

Thoughts?

pks-t replied:
Yes, I definitly agree, I've been reasoning about the same two options when I fixed this. I'd prefer to convert the function to return an `int` instead, as this is how nearly all of our functions which may fail operate. As such I deem it more idiomatic for this code base.

ethomson commented on src/refspec.c:
```diff
@@ -324,7 +324,7 @@ int git_refspec__dwim_one(git_vector *out, git_refspec *spec, git_vector *refs)
 			git_buf_clear(&buf);
 			if (git_buf_printf(&buf, formatters[j], spec->src) < 0)
-				return -1;
+				goto out_err;
```
Similar to the changes in `rebase.c`, we could have a `GITERR_CHECK_ALLOC_BUF` or something like that to check a `git_buf` for OOM status.  Or something.

pks-t replied:
Sounds reasonable and it would help the analysis as we'd be able to #nodef the macro and treat it as an abort when we run into allocation issues.

ethomson commented:
Thank you for this!  These sorts of things are super tough and annoying work but are amazingly beneficial.

pks-t commented:
New version of the PR. Changes:

* introduced `GITERR_CHECK_ALLOC_BUF`
* removed the check in `git_buf_joinpath`
* rewrote `rebase_alloc` to return an error code
* additional fixes for Coverity warnings

I'd like to highlight commit 0f1e2d2, which changes a field in `struct git_index_reuc_entry`. I couldn't think of any other way to reliably fix this issue, so I'm open to different proposals.

pks-t commented:
By the way, I'd like to wait for Coverity to check this through on my branch. I've currently exceeded the rate limit, will ping you when it's finished.

pks-t commented:
@ethomson Looks good to me, another three issues fixed and no new defects found.

ethomson commented:
Looks good to me.  Thanks for all the work here.
